pylint/extensions/typing.py

```diff
@@ -0,0 +1,306 @@
+        ),
+        "R6003": (
+            "Consider using alternative Union syntax instead of '%s'%s",
+            "consider-alternative-union-syntax",
+            "Emitted when 'typing.Union' or 'typing.Optional' is used "
+            "instead of the alternative Union syntax 'int | None'.",
+        ),
+    }
+    options = (
+        (
+            "py-version",
+            {
+                "default": (3, 7),
+                "type": "py_version",
+                "metavar": "<py_version>",
+                "help": (
+                    "Min Python version to use for typing related checks, "
+                    "e.g. ``3.7``. This should be equal to the min supported Python "
+                    "version of the project."
+                ),
+            },
+        ),
+        (
+            "runtime-typing",
+            {
+                "default": True,
+                "type": "yn",
+                "metavar": "<y_or_n>",
+                "help": (
+                    "Set to ``no`` if the app / libary does NOT need to "
+                    "support runtime introspection of type "
+                    "annotations. Only applies to Python version "
+                    "3.7 - 3.9"
+                ),
+            },
+        ),
+    )
+
+    def __init__(self, linter):
+        """Initialize checker instance."""
+        super().__init__(linter=linter)
+        self._alias_name_collisions: Set[str] = set()
+        self._consider_using_alias_msgs: List[DeprecatedTypingAliasMsg] = []
+
+    @lru_cache()
+    def _py37_plus(self):
+        return self.config.py_version >= (3, 7)
+
+    @lru_cache()
+    def _py39_plus(self):
+        return self.config.py_version >= (3, 9)
+
+    @lru_cache()
+    def _py310_plus(self):
+        return self.config.py_version >= (3, 10)
+
+    def _msg_postponed_eval_hint(self, node) -> str:
+        """Message hint if postponed evaluation isn't enabled."""
+        if self._py310_plus() or "annotations" in node.root().future_imports:
+            return ""
+        return ". Add 'from __future__ import annotations' as well"
+
+    @check_messages(
+        "deprecated-typing-alias",
+        "consider-using-alias",
+        "consider-alternative-union-syntax",
+    )
+    def visit_name(self, node: astroid.Name):
+        if (
+            self._py39_plus()
+            or self._py37_plus()
+            and self.config.runtime_typing is False
+        ) and node.name in ALIAS_NAMES:
+            self._check_for_typing_alias(node)
+        if (
+            self._py310_plus()
+            or self._py37_plus()
+            and self.config.runtime_typing is False
+        ) and node.name in UNION_NAMES:
+            self._check_for_alternative_union_syntax(node, node.name)
+
+    @check_messages(
+        "deprecated-typing-alias",
+        "consider-using-alias",
+        "consider-alternative-union-syntax",
+    )
+    def visit_attribute(self, node: astroid.Attribute):
+        if (
```

Comment:
We could create a cached function for that and the py3.10 test below it seems it's used in visit_name and visite_attribute at least.